File: Assets/src/model/enemies/IronJelly.cs
using System;
using UnityEngine;

[Serializable]
[ObjectInfo(description: "Invulnerable.\nAttacking the Iron Jelly pushes it away, first attacking any Creature in its way.")]
public class IronJelly : AIActor, IAnyDamageTakenModifier, IBodyTakeAttackDamageHandler {
  public IronJelly(Vector2Int pos) : base(pos) {
    faction = Faction.Neutral;
    hp = baseMaxHp = 99;
    ClearTasks();
  }

  public void HandleTakeAttackDamage(int damage, int hp, Actor source) {
    var isAdjacent = IsNextTo(source);
    var offset = source.pos - pos;
    var newPos = pos - offset;
    if (isAdjacent && offset != Vector2Int.zero && floor.InBounds(newPos)) {
      var pushedTile = floor.tiles[newPos];
      if (pushedTile.body != null) {
        Attack(pushedTile.body);
      }
      if (pushedTile.CanBeOccupied()) {
        pos = pushedTile.pos;
        Perform(new MoveBaseAction(this, pushedTile.pos));
      }
    }
  }

  internal override (int, int) BaseAttackDamage() => (99, 99);

  protected override ActorTask GetNextTask() {
    return new WaitTask(this, 1);
  }

  public int Modify(int input) {
    return 0;
  }
}

[Serializable]
class AttackOrMoveDirectionTask : TelegraphedTask {
  public AttackOrMoveDirectionTask(Actor actor, Vector2Int offset, int turns) :
    base(actor, turns, new AttackOrMoveDirectionBaseAction(actor, offset), ActionType.MOVE) {
    Offset = offset;
  }

  public Vector2Int Offset { get; }
}

[Serializable]
class AttackOrMoveDirectionBaseAction : BaseAction {
  public override ActionType Type => target == null ? ActionType.MOVE : ActionType.ATTACK;
  public Body target => actor.floor.bodies[actor.pos + offset];
  public readonly Vector2Int offset;

  public AttackOrMoveDirectionBaseAction(Actor actor, Vector2Int offset) : base(actor) {
    this.offset = offset;
  }

  public override void Perform() {
    if (target == null) {
      actor.pos = actor.pos + offset;
    } else {
      actor.AttackGround(actor.pos + offset);
    }
  }
}

File: web/src/model/enemies/IronJelly.ts
import { AIActor } from './AIActor';
import { ActorTask } from '../ActorTask';
import { WaitTask } from '../tasks/WaitTask';
import { MoveBaseAction } from '../BaseAction';
import {
  ANY_DAMAGE_TAKEN_MOD,
  type IAnyDamageTakenModifier,
} from '../../core/Modifiers';
import {
  BODY_TAKE_ATTACK_DAMAGE_HANDLER,
  type IBodyTakeAttackDamageHandler,
} from '../Body';
import { Faction } from '../../core/types';
import { Vector2Int } from '../../core/Vector2Int';
import { entityRegistry } from '../../generator/entityRegistry';
import type { Actor } from '../Actor';

/**
 * Invulnerable. Attacking the Iron Jelly pushes it away,
 * first attacking any Creature in its way.
 * Port of C# IronJelly.cs.
 */
export class IronJelly extends AIActor implements IAnyDamageTakenModifier, IBodyTakeAttackDamageHandler {
  readonly [ANY_DAMAGE_TAKEN_MOD] = true as const;
  readonly [BODY_TAKE_ATTACK_DAMAGE_HANDLER] = true as const;

  constructor(pos: Vector2Int) {
    super(pos);
    this.faction = Faction.Neutral;
    this.hp = this._baseMaxHp = 99;
    this.clearTasks();
  }

  baseAttackDamage(): [number, number] {
    return [99, 99];
  }

  /** IAnyDamageTakenModifier — reduce all damage to 0 (invulnerable). */
  modify(_input: number): number {
    return 0;
  }

  /** When attacked, get pushed away and attack anything in the pushed-to tile. */
  handleTakeAttackDamage(_damage: number, _hp: number, source: Actor): void {
    if (!this.floor) return;
    const isAdjacent = this.isNextTo(source);
    const offset = Vector2Int.sub(source.pos, this.pos);
    const newPos = Vector2Int.sub(this.pos, offset);

    if (isAdjacent && !Vector2Int.equals(offset, Vector2Int.zero) && this.floor.inBounds(newPos)) {
      const tile = this.floor.tiles.get(newPos);
      if (!tile) return;

      // Attack whatever is in the way
      if (tile.body) {
        this.attack(tile.body as import('../Body').Body);
      }

      // Move to the new position if it's free
      if (tile.canBeOccupied()) {
        this.pos = tile.pos;
        this.perform(new MoveBaseAction(this, tile.pos));
      }
    }
  }

  protected getNextTask(): ActorTask {
    return new WaitTask(this, 1);
  }
}

entityRegistry.register('IronJelly', IronJelly);
